<!-- XML -->
<launch>
    <arg name="width" default="1280"/>
    <arg name="height" default="720"/>

    <!-- FROM v4l_imx390_raw.launch -->

    <!-- You can check the device ID and subdev ID for the IMX390 camera attached
    by running /opt/edge_ai_apps/scripts/setup_cameras.sh.
    Accordingly please update the parameters pass as arguments -->
    <arg name="device" default="/dev/imx390-1-2"/>
    <arg name="subdev" default="/dev/imx390-1-2-subdev"/>
    <!-- framerate (integer): inactive, 30 FPS (default, max) -->
    <!-- Can be enabled by e.g., adding videorate before tiovxdlcolorconvert, but it creases A72 loading -->
    <arg name="framerate" default="30"/>
    <!-- node name -->
    <arg name="node_name" default="gscam_node"/>
    <!-- camera name: also used as namespace for the output topic name -->
    <arg name="camera_name" default="imx390"/>
    <!-- DCC VISS binary file -->
    <arg name="dcc_isp_file" default="/opt/imaging/imx390/dcc_viss.bin"/>
    <!-- DCC 2A binary file -->
    <arg name="dcc_2a_file" default="/opt/imaging/imx390/dcc_2a.bin"/>
    <!-- LDC binary file -->
    <!-- <arg name="ldc_dcc_file" default="/opt/imaging/imx390/imx390_35244_equidistant_1280x720_LUT.bin"/>-->
    <arg name="ldc_dcc_file" default="/opt/imaging/imx390/imx390_robotcam4_equidistant_1280x720_LUT.bin"/>
    <!-- <arg name="ldc_dcc_file" default="/opt/imaging/imx390/robotcam3_equidistant_1280x720_LUT.bin"/>-->
    <!-- camera_info URL: replace with camera_info from camera calibration -->
    <!-- <arg name="camera_info_url" default="package://gscam/config/IMX390_HD_camera_info.yaml"/> -->
    <!-- <arg name="camera_info_url" default="file:///opt/imaging/imx390/imx390_35244_fisheye.yml"/> -->
    <!-- NOTE: Using the rectified intrinsics here due to the tiovxldc being integrated into the gst pipeline -->
    <!--<arg name="camera_info_url" default="file:///opt/imaging/imx390/imx390_35244_equidistant_1280x720_rect.yml"/>-->
    <!--<arg name="camera_info_url" default="file:///opt/imaging/imx390/robotcam3_equidistant_1280x720_rect.yml"/>-->
    <arg name="camera_info_url" default="file:///opt/imaging/imx390/imx390_robotcam4_equidistant_1280x720_rect.yml"/>

    <!-- GStreamer pipeline specified in gscam_config was tested with IMX390 camera,
        'edgeai-tiovx-modules' and 'edgeai-gst-plugins' are assumed to be already installed in TDA4 ROS container.
        Raw resolution is 1936 x 1096 at 30 FPS.
        Note: /dev/v4l-subdev ID can change, depending on the device attached, and at reboot. -->
    <node pkg="gscam" name="$(arg node_name)" type="gscam" output="screen">
        <param name="gscam_config"
            value="v4l2src device=$(arg device) do-timestamp=true ! videorate! video/x-bayer, width=1936, height=1096, format=rggb12, framerate=1/1 !
            tiovxisp sink_0::device=$(arg subdev) sensor-name=SENSOR_SONY_IMX390_UB953_D3 dcc-isp-file=$(arg dcc_isp_file)
            sink_0::dcc-2a-file=$(arg dcc_2a_file) format-msb=11 ! video/x-raw, format=NV12 !
            tiovxldc sensor-name=SENSOR_SONY_IMX390_UB953_D3 lut-file=$(arg ldc_dcc_file) ldc-ds-factor=2 ldc-table-width=$(arg width) ldc-table-height=$(arg height) !
            video/x-raw, format=NV12, width=$(arg width), height=$(arg height) !
            tiovxdlcolorconvert target=1 out-pool-size=4 "/>
        <param name="camera_name"     value="$(arg camera_name)"/>
        <param name="camera_info_url" value="$(arg camera_info_url)"/>
        <param name="width"           value="$(arg width)"/>
        <param name="height"          value="$(arg height)"/>
        <param name="framerate"       value="$(arg framerate)"/>
        <param name="sync_sink"       value="false"/>
        <param name="use_gst_timestamps" value="false"/>
        <!-- image encoding: "yuv420" - publish in NV12 -->
        <param name="image_encoding"  value="yuv420"/>
    </node>

    <!-- color conversion for input image_raw for visualization -->
    <node pkg="ti_viz_nodes" type="viz_color_conv_yuv2rgb" name="viz_color_conv_yuv2rgb_node" output="screen" args="" required="true">
        <param name="width"             value="$(arg width)"/>
        <param name="height"            value="$(arg height)"/>
        <param name="input_yuv_topic"   value="imx390/image_raw"/>
        <param name="output_rgb_topic"  value="imx390/image_raw_rgb"/>
        <param name="yuv_format"        value="YUV420"/>
        <param name="yuv420_luma_only"  value="true"/>
    </node>

    <node pkg="d3_motorctl" type="apriltag_odom.py" name="apriltag_node" output="screen">
        <param name="camera_name" value="imx390"/>
        <remap from="/image_raw" to="/imx390/image_rgb"/>
    </node>

    <node pkg="tf" type="static_transform_publisher" name="map_to_apriltag21" args="0 0 2.47 0 0 3.14159 map apriltag21 100" />
    <!-- position is correct, orientation of base is wrong <node pkg="tf" type="static_transform_publisher" name="map_to_apriltag420" args="1.829 0 0 4.7124 0 1.5708 map apriltag420 100" /> -->
    <node pkg="tf" type="static_transform_publisher" name="map_to_apriltag420" args="1.829 0 0 4.7124 0 1.5708 map apriltag420 100" />
    <!-- [ACTIVE] rear camera pointed up, aligned per ROS REP 103 --> <node pkg="tf" type="static_transform_publisher" name="imx390_rear_optical_to_base_link" args="0 0.125 0 1.57079 0 0 imx390_rear_optical base_link 100" />
    <!-- rear camera pointed back, aligned with base_link <node pkg="tf" type="static_transform_publisher" name="imx390_rear_optical_to_base_link" args="0 0.125 0 0 0 0 imx390_rear_optical base_link 100" /> -->
    <!-- rear camera pointed back, aligned per ROS REP 103 <node pkg="tf" type="static_transform_publisher" name="imx390_rear_optical_to_base_link" args="0 0 -0.125 0 1.5708 1.5708 imx390_rear_optical base_link 100" /> -->

</launch>

